Clean up Cl2DecodeFrm3, Cl2DecodeFrm5, Cl2DecodeLightTbl & Cl2DecodeFrm6

AJenbo opened the pull request:
This solves some of the diff en Cl2DecodeFrm3 amd Cl2DecodeFrm5 (they are identical except for what function they call in turn).

![image](https://user-images.githubusercontent.com/204594/59231540-4ac77980-8be1-11e9-8c25-53a402c4694e.png)

The remaning issue is that `shl exa 0x4` comes before 
```
sub esi, [ebp+0x08]
add edi, [ebp+0x08]
```

`shl exa 0x4` appears to relate to `PitchTbl[sy - 16 * CelSkip]`

```
sub esi, [ebp+0x08]
add edi, [ebp+0x08]
```
comes from
```
	nDataSize -= nDataStart;
	pRLEBytes += nDataStart;
```

so why they appear swapped in ASM I don't know, and less so how to try and correct it :/

AJenbo commented:
Looking at the PowerPC version there is one interesting difference, instead of `PitchTbl[sy - 16 * CelSkip]` it shows up as `PitchTbl[sy + CelSkip * -16]`, but it doesn't appear to make a difference.

AJenbo commented:
Could the issue be that this was an inline function?
https://github.com/AJenbo/deviled-apples/blob/master/Ghidra/Source/engine.cpp#L3156

```
longlong FUN_1003ce1c(longlong param_1, int param_2, int param_3, int param_4, int *param_5)

{
	undefined *puVar3;
	ulonglong uVar5;
	longlong lVar6;

	uVar5 = (byte *)((int)param_1 + nCel * 4);
	nDataSize = param_1 + uVar5;
	puVar3 = (undefined *)(nDataSize + CelSkip);
	sVar2 = CONCAT11(puVar3[1], *puVar3);
	if ((longlong)sVar2 != 0) {
		if (CelCap == 8) {
			nDataSize = 0;
		} else {
			puVar3 = (undefined *)(nDataSize + CelCap);
			nDataSize = (int)CONCAT11(puVar3[1], *puVar3);
		}
		if (nDataSize == 0) {
			nDataSize = (byte *)((int)param_1 + (nCel + 1) * 4);
		}
		*param_5 = nDataSize - (int)sVar2;
		return nDataSize + (longlong)sVar2;
	}
	return 0;
}
```

galaxyhaxz commented:
You added assert strings that aren't there in the debug release.
Also, given the MAC release had to rewrite all parts of ASM, the code is a bit different. The endianness is flipped so it has to call a separate function to flip it back. The `render.cpp` is entirely different and uses a jump table, which is likely because the original was asm and used a patch table for the switch.

AJenbo commented:
Removed the asserts, I had added them so that the various functions where easier to compare.

Don't you think that it's still likely that PC and Mac versions shared code base for this and only the endian flipping and compiler (maybe the one used on Mac didn't support inlining) where different?

galaxyhaxz commented:
How would they share a code base? Mac uses PPC and windows x86 ASM which are entirely different. Looking at render.cpp, it's definitely nothing like the windows version. I think most likely the main team only worked on the windows version hence why it took several years for the updated Mac version to come out. I think a separate team would have forked and made the necessary changes. This is also likely the case for Diablo II as well.

AJenbo commented:
I agree that render.cpp and other parts written in ASM had to be completely different, but the part of engine.cpp that we are dealing with here looks to have been written in C++. So for it something like this should work on both plat forms:

```CPP
#ifdef __MAC__
#define SwapLE(x) (((x & 0x000000FF) << 24) | ((x & 0x0000FF00) << 8) | ((x & 0x00FF0000) >> 8) | ((x & 0xFF000000) >> 24))
#else
#define SwapLE(x) (x)
#endif

void __inline CalcCelDataSize(BYTE *pCelBuff, int nCel, int CelSkip, int CelCap)
{
	int nDataStart, nDataSize;
	BYTE *pRLEBytes;
	DWORD *pFrameTable;

	pFrameTable = (DWORD *)pCelBuff;
	pRLEBytes = &pCelBuff[pFrameTable[nCel]];
	nDataStart = *(WORD *)&pRLEBytes[CelSkip];
	nDataStart = SwapLE16(nDataStart);
	if (!nDataStart)
		return 0;

	if (CelCap == 8)
		nDataSize = pFrameTable[nCel + 1] - pFrameTable[nCel];
	else
		nDataSize = *(WORD *)&pRLEBytes[CelCap];
	nDataStart = SwapLE16(nDataStart);

	return nDataSize - nDataStart;
}
```

My understanding is that they start working on the PSX or Mac port circa 1 year after the games release. The PSX port was a hard fork with no code sharing. But for the Mac port a person was hired to work on it in-house and the code base was shared so that you could compile either targets from the same source.

Predelnik commented:
So the following solves it:
* Make new variable (e.g. named `size`)
* Assign `size` to `nDataSize` directly after `pDecodeTo` assignment.
* Use this new `size` in `Cl2DecDatLightTbl1` etc.

AJenbo commented:
Applying @Predelnik's suggestion and adjusting it to what's in the Mac port this also solved Cl2DecodeLightTbl.

mewmew commented:
> Applying @Predelnik's suggestion and adjusting it to what's in the Mac port this also solved Cl2DecodeLightTbl and Cl2DecodeFrm6.

Yay!
